crates/runtime/src/libcalls.rs

```diff
@@ -137,3 +142,195 @@ pub unsafe extern "C" fn wasmtime_imported_memory32_size(
 
     instance.imported_memory_size(memory_index)
 }
+
+/// Implementation of `table.copy` when both tables are locally defined.
+#[no_mangle]
+pub unsafe extern "C" fn wasmtime_table_copy_defined_defined(
+    vmctx: *mut VMContext,
+    dst_table_index: u32,
+    src_table_index: u32,
+    dst: u32,
+    src: u32,
+    len: u32,
+    source_loc: u32,
+) {
+    let dst_table_index = DefinedTableIndex::from_u32(dst_table_index);
+    let src_table_index = DefinedTableIndex::from_u32(src_table_index);
+    let source_loc = ir::SourceLoc::new(source_loc);
+    let instance = (&mut *vmctx).instance();
+    let dst_table = instance.get_defined_table(dst_table_index);
+    let src_table = instance.get_defined_table(src_table_index);
+    if let Err(trap) = Table::copy(dst_table, src_table, dst, src, len, source_loc) {
+        raise_lib_trap(trap);
```

Comment:
This ends up not running destructors and is pretty dangerous. For example I think `instance` above may be leaked here, preventing the instance from being dropped? In any case could this be reorganized to:

```rust
let result = {
    // ...
};
if let Err(trap) = result {
    raise_lib_trap(trap);
}
```

with a comment about how the block-structure is required to make sure dtors run?

Reply:
So a possible alternative to this as well might be to do something entirely different wrt traps. For example each of these intrinsic functions could return an i32 which, if zero, would cause a trap. The generated code would then do the test-and-trap itself and have an entry in the trap map for why the trap is happening at that address.

I feel like that may be a bit cleaner than doing the longjmp ourselves, but functionally it's all the same so it doesn't matter too too much. I'm just overly worried about these unwinding functions :(

Reply:
Yeah, we would probably want to make `TrapCode` be `repr(C)` and also have a `Ok` variant or something like that for success (or maybe a new `ResultCode` so that `TrapCode` itself doesn't have an `Ok`).

This seems like a good plan, but also something we can do not in this particular PR.